k2/csrc/array.h

```diff
@@ -410,10 +417,46 @@ class Array2 {
     that
      it will have a different memory layout than the input.
   */
-  Array2<T> To(ContextPtr ctx) {
-    // TODO
-    Array2<T> array;
-    return array;
+  Array2<T> To(ContextPtr ctx) const {
+    if (dim0_ * dim1_ == 0) {
+      if (!region_) return *this;
```

Comment:
I think there should be no this line. As the parameter `ctx` is valid, we should always returen a valid `Array2` to users, even if the size is `0` and `this->region_ == nullptr`

Reply:
We could perhaps require that region_ never be nullptr.  In fact that may already be a requirement.
Notice that NewRegion() allows zero bytes to be requested.

Reply:
Well, sometimes we need a default constructor for Array, such as `Array1(): dim_(0), region_(nullptr)`. This has been used in a few algorithms written by you before. If we do not allow such a constructor, we need to make sure we always have a `Context` in all usage of Array, to call `NewRegion()`. Which one do you prefer?

Reply:
This follows the following comment:
https://github.com/danpovey/k2/blob/2a847823668699959e0be10687acb077a5c8b58c/k2/csrc/array.h#L401-L403

I think it should be an error for an empty Array2 to invoke the `To` method.

Reply:
e.g. `RaggedShapeDim` which contains `Array1` using a default constructor

Reply:
I guess the above documentation is something old and should be updated.

Reply:
> Well, sometimes we need a default constructor for Array, such as `Array1(): dim_(0), region_(nullptr)`. This has been used in a few algorithms written by you before. If we do not allow such a constructor, we need to make sure we always have a `Context` in all usage of Array, to call `NewRegion()`. Which one do you prefer?

We will consider that an Array1/Array2 constructed with default constructor is "not valid" and only assignment may be performed; other operations may segfault.

Reply:
mm yes, remove that old documentation.

Reply:
OK, seems we need to do `K2_CHECK(region_ != nullptr)`  in all algorithms for `Array1` and `Array2`

Reply:
old documentation is removed.

The function will segfault when invoked with region_ == nullptr.